x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/filestructurefinder/FileStructure.java

```diff
@@ -150,7 +152,8 @@ public String toString() {
     private final Character quote;
     private final Boolean shouldTrimFields;
     private final String grokPattern;
-    private final List<String> timestampFormats;
+    private final List<String> jodaTimestampFormats;
```

Comment:
Reading through the code, I wonder if it would make things simpler to encapsulate this in a `TimestampFormat` class which will then own the joda/java pairs (or additional types if the future brings them). That would create that sort of 1-1 mapping between the alternatives. Just a thought

Reply:
I'm not convinced that at the level of individual timestamp formats we want the Joda/Java formats grouped into pairs.

The output of the endpoint can currently contain something like this:

```
  "joda_timestamp_formats" : [
    "MMM dd YYYY HH:mm:ss", "MMM  d YYYY HH:mm:ss"
  ],
  "java_timestamp_formats" : [
    "MMM dd yyyy HH:mm:ss", "MMM  d yyyy HH:mm:ss"
  ],
```

As a consumer of this information you'd want either the Joda formats or the Java formats, never both.  Suppose the endpoint returned something like this:

```
  "timestamp_formats" : [
    {
      "joda" : "MMM dd YYYY HH:mm:ss",
      "java" : "MMM dd yyyy HH:mm:ss"
    },
    {
      "joda" : "MMM  d YYYY HH:mm:ss",
      "java" : "MMM  d yyyy HH:mm:ss"
    }
```

That would be very annoying to use for somebody who wanted both the Joda formats _or_ both the Java formats.  All consumers would end up having to write their own transformation function to get back to a list of formats of the same type.

So the other way to avoid the doubled method calls in a few places would be to have a `TimestampFormats` (plural) container that contains both lists of formats.  The problem I foresee with this is that eventually we'll want to completely remove the Joda formats, so then the class will only have one member and will just be an annoyance.

Reply:
I see your point. Sounds good!